### PR TITLE
Update to gromacs 2021

### PR DESCRIPTION
**Pull request by longr**
Describe your pull request here

* Updated README for current versions
* Removed 2018 and 2019 recipes as those branches are now out of maintenance upstream.
* Added 2021 subdirectory for 2021 version.

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

For members of the Bioconda project, the following command is also available:

<table>
  <tr>
    <td><code>@BiocondaBot please merge</code></td>
    <td>Upload built packages/containers and merge a PR. <br />Someone must approve a PR first! <br />This reduces CI build time by reusing built artifacts.</td>
  </tr>
</table>

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

**Comment by douglowe** [comment]
It looks like the problem *might* be that the OSX deployment target is 10.9, but there are features in the code which were introduced in 10.13, e.g.:
```
06:23:23 BIOCONDA INFO (ERR) /Users/distiller/project/miniconda/conda-bld/gromacs_1615299449709/work/src/gromacs/modularsimulator/computeglobalselement.cpp:393:62: error: 'value' is unavailable: introduced in macOS 10.13
06:23:23 BIOCONDA INFO (ERR)         return std::any_cast<ISimulatorElement*>(cachedValue.value());
``` 

This conda-forge script sets the MACOS_DEPLOYMENT_TARGET flag, with a default of 10.9: [run_conda_forge_build_setup_osx](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/a1026adb523b6562c16329170e7e304a25ed4033/recipe/run_conda_forge_build_setup_osx)

In conda-forge recipes it is possible to change this default, using a file such as [.ci_support/osx_64_.yaml](https://github.com/conda-forge/prismatic_split-feedstock/blob/master/.ci_support/osx_64_.yaml).

Perhaps it is possible to do similar for a bioconda recipe?

**Comment by douglowe** [comment]
@mabraham - would you have any suggestions on how this might be done?

**Review by mabraham — COMMENTED** [review]
No real idea (I'm a GROMACS expert, not a bioconda one) but grepping the source code suggests https://github.com/bioconda/bioconda-recipes/blob/master/recipes/razers3/build.sh

**Comment by longr** [comment]
> No real idea (I'm a GROMACS expert, not a bioconda one) but grepping the source code suggests https://github.com/bioconda/bioconda-recipes/blob/master/recipes/razers3/build.sh

Thanks, will give this a try.  Googled a few things, but did not think to grep the src.

File: recipes/gromacs/2021/build.sh [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+for ARCH in SSE2 AVX_256 AVX2_256 AVX_512; do
+  cmake_args=(
+    -DSHARED_LIBS_DEFAULT=ON
+    -DBUILD_SHARED_LIBS=ON
+    -DGMX_PREFER_STATIC_LIBS=NO
+    -DGMX_BUILD_OWN_FFTW=OFF
+    -DGMX_DEFAULT_SUFFIX=ON
+    -DREGRESSIONTEST_DOWNLOAD=ON
+    -DGMX_GPU=OpenCL
+    -DGMX_USE_OPENCL=ON
```
[the remark this answers is not in the document]
Comment (longr) [reply recipes/gromacs/2021/build.sh:15]: Thanks.  I am not sure how GitHub comments work, so forgive the ignorance, but are you suggesting I remove line 15 or lines 12 - 15.

[the remark this answers is not in the document]
Comment (mabraham) [reply recipes/gromacs/2021/build.sh:22]: line 15 only

[the remark this answers is not in the document]
Comment (longr) [reply recipes/gromacs/2021/build.sh:22]: Thanks.  Do you know much about the changes to line 14?  It was previously set to =ON, but this breaks on the 2021 build so I changed it to OpenCL, I could not find anything in the release notes about this.  Not sure if I picked the correct / best option.

[the remark this answers is not in the document]
Comment (mabraham) [reply recipes/gromacs/2021/build.sh:22]: Yes, OpenCL is correct. SOrry the release notes missed it!

[the remark this answers is not in the document]
Comment (longr) [reply recipes/gromacs/2021/build.sh:22]: It is fine, I may have missed it.

File: recipes/gromacs/2021/build.sh [hunk 2]
```diff
@@ -0,0 +1,144 @@
+#!/bin/bash
+
+# gromacs > 2021 requires a non default OSX version
```
Comment (mabraham) [inline recipes/gromacs/2021/build.sh:3]: Suggest "C++17 is not properly supported on early OSX versions"

**Comment by longr** [comment]
@mabraham Any thoughts on what is causing the OSX build error.  @douglowe and I have been scratching our heads on it and cannot figure it out.

**Comment by douglowe** [comment]
There is an error about CTL_KERN being undefined (in hardware/device_management_ocl.cpp). Not 100% certain it's causing this build error, but should this file include `#include <sys/sysctl.h>` so that this environmental variable is available anyway?

**Comment by mabraham** [comment]
> There is an error about CTL_KERN being undefined (in hardware/device_management_ocl.cpp). Not 100% certain it's causing this build error, but should this file include `#include <sys/sysctl.h>` so that this environmental variable is available anyway?

I think it should, and presumably one of the other headers was supplying it until one of them got changed. I can get this fixed upstream, but it would be good to know the fix will work. Can we introduce a patch to that file before the build?

Alternatively, don't support OpenCL for 2021.1

**Comment by longr** [comment]
> > There is an error about CTL_KERN being undefined (in hardware/device_management_ocl.cpp). Not 100% certain it's causing this build error, but should this file include `#include <sys/sysctl.h>` so that this environmental variable is available anyway?
> 
> I think it should, and presumably one of the other headers was supplying it until one of them got changed. I can get this fixed upstream, but it would be good to know the fix will work. Can we introduce a patch to that file before the build?
> 
> Alternatively, don't support OpenCL for 2021.1

I managed to patch this, and it now builds. Looks like this was the issue (also, this is version 2021, not looked to see if it is fixed in 2021.1)

**Comment by longr** [comment]
@BiocondaBot please add label

**Review by dpryan79 — APPROVED** [review]

**Comment by dpryan79** [comment]
@bioconda-bot please merge

**Comment by mabraham** [comment]
> > > There is an error about CTL_KERN being undefined (in hardware/device_management_ocl.cpp). Not 100% certain it's causing this build error, but should this file include `#include <sys/sysctl.h>` so that this environmental variable is available anyway?
> > 
> > 
> > I think it should, and presumably one of the other headers was supplying it until one of them got changed. I can get this fixed upstream, but it would be good to know the fix will work. Can we introduce a patch to that file before the build?
> > Alternatively, don't support OpenCL for 2021.1
> 
> I managed to patch this, and it now builds. Looks like this was the issue (also, this is version 2021, not looked to see if it is fixed in 2021.1)

It isn't

**Comment by mabraham** [comment]
> > > > There is an error about CTL_KERN being undefined (in hardware/device_management_ocl.cpp). Not 100% certain it's causing this build error, but should this file include `#include <sys/sysctl.h>` so that this environmental variable is available anyway?
> > > 
> > > 
> > > I think it should, and presumably one of the other headers was supplying it until one of them got changed. I can get this fixed upstream, but it would be good to know the fix will work. Can we introduce a patch to that file before the build?
> > > Alternatively, don't support OpenCL for 2021.1
> > 
> > 
> > I managed to patch this, and it now builds. Looks like this was the issue (also, this is version 2021, not looked to see if it is fixed in 2021.1)
> 
> It isn't

Upstream MR at https://gitlab.com/gromacs/gromacs/-/merge_requests/1284 targeting 2021.2

**Pull request merged**